contracts/bid/src/hosted/transaction.rs

```diff
@@ -0,0 +1,286 @@
+
+use crate::fake_abi;
+use crate::{contract_constants::*, leaf::BidLeaf, Contract};
+use alloc::vec::Vec;
+use canonical::Store;
+use core::ops::DerefMut;
+use dusk_blindbid::Bid;
+use dusk_bls12_381::BlsScalar;
+use dusk_jubjub::JubJubAffine;
+use dusk_pki::{Ownable, PublicKey};
+use phoenix_core::Note;
+use schnorr::Signature;
+
+impl<S: Store> Contract<S> {
+    /// This function allows to the contract caller to setup a Bid related to a
+    /// one-time identity of his/her property that will allow the user to
+    /// participate in the bidding process of the consensus as well as to
+    /// prove that is part of the bidders set.
+    ///
+    /// This function will first of all, verify that the Bid is correct by
+    /// verifying the BidCorrectness Proof.
+    /// Then it will include the Bid into the PoseidonTree of the contract and
+    /// link the One-Time identity of the user to the index that the bid
+    /// occupies in the tree. Finally it will execute an inter-contract call
+    /// sending the `spending_proof` and a `note` to the transfer contract.
+    /// Which will execute the transaction of Dusk to the contract account.
+    pub fn bid(
+        &mut self,
+        mut bid: Bid,
+        correctness_proof: Vec<u8>,
+        _spending_proof: Vec<u8>,
+    ) -> (bool, usize) {
+        // Setup error flag to false
+        let mut err_flag = false;
+
+        // Verify proof of Correctness of the Bid.
+        if !fake_abi::verify_proof(
+            correctness_proof,
+            crate::BID_CORRECTNESS_VK.to_vec(),
+            b"bid-correctness".to_vec(),
+            PublicInput::AffinePoint(bid.commitment(), 0, 0)
+                .to_bytes()
+                .to_vec(),
+        ) {
+            err_flag = true;
+            return (err_flag, usize::MAX);
+        }
+
+        // Obtain the current block_height.
+        let block_height = dusk_abi::block_height();
+        // Compute maturity & expiration periods
+        let expiration = block_height + MATURITY_PERIOD + EXPIRATION_PERIOD;
+        let eligibility = block_height + MATURITY_PERIOD;
+
+        // Mutate the Bid and add the correct timestamps.
+        bid.set_eligibility(eligibility);
+        // FIXME: This should not be needed. We should have a better API in blindbid.
+        //  Since the API for blindbid was decided to be set as `extend_expiration` instead of
+        // `set_expiration`. We now are forced to ensure that the expiration is 0 here and then, we
+        // sum `expiration` to it.
+        assert!(bid.expiration() == 0u64);
+        bid.extend_expiration(expiration);
+
+        // Panic and stop the execution if the same one-time-key tries to
+        // bid more than one time.
+        let idx = if self
+            .key_idx_map()
+            // If no entries are found for this PK, add it to the map and the
+            // tree
+            .get(*bid.stealth_address().pk_r())
+            .unwrap()
```

Comment:
Shouldn't we return `false` instead of `unwrap`?

Reply:
Why? If there are no entries. It means that there;'s no bidder with this Pk. Therefore we need to add the bid to the tree and the bidder pk to the map.

On any other case, we need to err.